### PR TITLE
OKTA-521284 - Updating "Back to sign in" link logic on Terminal pages to match v2

### PR DESCRIPTION
**Pull request by glenfannin-okta**
## Description:

The purpose of this PR is to fix the "Back to sign in" link on various terminal pages.

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-521284](https://oktainc.atlassian.net/browse/OKTA-521284)

### CODE REVIEW

File: src/v3/src/hooks/useOnSubmit.ts [hunk 1]
```diff
@@ -51,6 +53,15 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
       stepToRender,
     } = options;
 
+    let { stateHandle } = currTransaction!.context;
+
+    if (step === 'cancel') {
+      authClient?.transactionManager.clear({ clearIdxResponse: false });
```
Comment (glenfannin-okta) [inline src/v3/src/hooks/useOnSubmit.ts:59]: I added this line to match what v2 is currently doing when it encounters a `cancel` action: https://github.com/okta/okta-signin-widget/blob/master/src/v2/controllers/FormController.ts#L165-L177

File: src/v3/src/hooks/useOnSubmit.ts [hunk 2]
```diff
@@ -51,6 +53,15 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
...
+      if (previousTransaction) {
+        stateHandle = previousTransaction.context.stateHandle;
+      }
```
Comment (glenfannin-okta) [inline src/v3/src/hooks/useOnSubmit.ts:62]: To explain why this fixes the issue: 

This is ultimately an auth-js issue, and we are adding in a temporary workaround to prevent broken links (this issue exists in v2 as well). 
[This](https://github.com/okta/okta-auth-js/blob/master/lib/idx/run.ts#L130) is the entrypoint in auth-js. If we follow the flow, you can see it eventually hits [this](https://github.com/okta/okta-auth-js/blob/master/lib/idx/introspect.ts#L30) line, which finally lands [here](https://github.com/okta/okta-auth-js/blob/master/lib/TransactionManager.ts#L340). This is problematic because it is expecting the stateHandle to match the stateHandle from the last request. However, we are sending the stateHandle from the current transaction (which was updated on the terminal response). When they do not match, this line returns null and causes `introspect` to be called instead of the `cancel` action. Calling introspect after a failure, just results in the same failure, causing this link to appear broken. The only way the user can escape this is by refreshing the page.

Comment (glenfannin-okta) [reply src/v3/src/hooks/useOnSubmit.ts:62]: Ticket opened to auth-js team: https://oktainc.atlassian.net/browse/OKTA-528448

Comment (shawyu-okta) [reply src/v3/src/hooks/useOnSubmit.ts:62]: Attach a short explanation and ticket to the top of this block for context and also to discourage anyone from adding code inside of this logic for any reason

File: src/v3/src/hooks/useOnSubmit.ts [hunk 3]
```diff
@@ -91,12 +102,13 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
     }
     payload = toNestedObject(payload);
     if (currTransaction!.context.stateHandle) {
-      payload.stateHandle = currTransaction!.context.stateHandle;
+      payload.stateHandle = stateHandle;
     }
```
Comment (shawyu-okta) [inline src/v3/src/hooks/useOnSubmit.ts:106]: If we're going to be able to remove the logic above, would it be easier to leave this untouched even if it's a bit more verbose? i.e. would be nice to be able to just remove the earlier code when the auth-js issue is resolved without having to undo this change.

**Review by shawyu-okta — APPROVED** [review]
small comments but overall lgtm

File: src/v3/src/hooks/useOnSubmit.ts (hunk 3 above)
Comment (glenfannin-okta) [reply src/v3/src/hooks/useOnSubmit.ts:106]: Totally, it would require me to move that if block below this if block (to overwrite the stateHandle that is set in the payload). But as you mentioned, once that code can be removed, it can easily be deleted without modifying this line as well. 

Is that what you meant?

Comment (shawyu-okta) [reply src/v3/src/hooks/useOnSubmit.ts:106]: Oh I forgot that you'd have to move the block down. If it doesn't affect the functionality nor cause any additional complication in the code then I think that would be desirable. If it makes things difficult then it's not a big deal at all to leave things as-is.

File: src/v3/src/components/Widget/index.tsx [hunk 4]
```diff
@@ -82,6 +82,7 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
   });
   const [message, setMessage] = useState<IdxMessage | undefined>();
   const [idxTransaction, setIdxTransaction] = useState<IdxTransaction | undefined>();
+  const [previousTransaction, setPreviousTransaction] = useState<IdxTransaction | undefined>();
```
Comment (shuowu-okta) [inline src/v3/src/components/Widget/index.tsx:85]: you don't need this state, see line 88

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: Oh yea! Great catch, I actually tried to use this initially, I think it may have something to do with when it gets updated but this previous transaction did not have the correct stateHandle. The stateHandle in this object matched that from the current transaction, which is why I created a new state object. 

When I saw that I didnt debug more into why, mostly because I didnt want to make any changes that would negatively impact the code that is using this ref.

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: Here is what I was talking about: 
image

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: Ref is the existing ref that was created, current transaction is the current transaction, and prev is the new state object I created. 

My fear was changing something with that prev transaction ref and causing breakage for whatever was using it. Especially if this logic is going to go away after a bug fix, I didn't think it would be worth it to put it a bunch of effort into that.

Comment (shuowu-okta) [reply src/v3/src/components/Widget/index.tsx:85]: the last one in your log (probably what you tracked in state), does not look like a stateHandle, but more like a token instead, what you get if you decode it?

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: This is what I get when I decode it, it is directly from the `stateHandle` property that is in the transaction context. 

It is 100% the difference of which transaction is being used. The ref uses the last successful transaction whereas this state object uses the "currentTransaction". Maybe the name is confusing. I think I should just create a stateProp called `savedStateHandle` or `stateHandle` or something like that to avoid confusion. 

image

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: You can see here, this is the transaction we are using the stateHandle from for this workaround: https://github.com/okta/okta-signin-widget/pull/2752/files#diff-114b8aded3d172ffc719a9452df5b38cfe8f9e82362d922bc0ca4f7f4440005aR111

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: Where as the ref uses the last successful transaction. So maybe the best idea is to just save the stateHandle as a state prop and not the entire transaction.

Comment (shuowu-okta) [reply src/v3/src/components/Widget/index.tsx:85]: ohh, this one is for the terminal state issue, authjs clears transaction related data, I don't think a workaround can be applied in widget side. I was suggesting https://github.com/okta/okta-signin-widget/blob/21b98876539d6c48a7108a44de337027085f41d7/src/v2/view-builder/utils/LinksUtil.js#L121

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: Ahh, so the above solution does work, see the integration test, (albeit not the greatest). I took at look at that `getBackToSignInLink` but ultimately didnt do it because in this scenario, the cancel action exists on the transaction. How will we know when to use it? Or are you saying to always use this logic regardless if `cancel` action exists or not?

Comment (shuowu-okta) [reply src/v3/src/components/Widget/index.tsx:85]: As I said before, it's how v2 works, and authjs currently has issue for this scenario. So I would suggest to follow v2 instead of hacking another solution.

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: Yep, I recall :D but this is how v2 handles it today: https://github.com/okta/okta-signin-widget/blob/21b98876539d6c48a7108a44de337027085f41d7/src/v2/view-builder/views/TerminalView.js#L158-L164

This only happens if the cancel object doesnt exist, in this scenario, the cancel action is present. So are you saying to always use this logic even if the cancel action is present?

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: I am not against doing this, I just want to make sure this doesnt cause any other issues by not using the cancel action when it is expected to be used.

Comment (shuowu) [reply src/v3/src/components/Widget/index.tsx:85]: v3 should work if v2 works, which does not send cancel request in the current version (based on what i read from code)

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: V2 is broken in this scenario (and other scenarios like this) as well (same exact issue, it calls introspect instead of cancel) 😅

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:85]: The only fix would be to always use this `getBackToSignInLink` logic in v3 until a fix can be applied to auth-js. Or what I have above with the stateHandle. 

Which method is preferred?

File: src/v3/src/util/settingsUtils.ts [hunk 5]
```diff
@@ -73,3 +73,25 @@ export const getLanguageCode = (widgetProps: WidgetProps): LanguageCode => {
 
   return (supportedLanguages[supportedPos] ?? config.defaultLanguage) as LanguageCode;
 };
+
+export const getBackToSignInUri = (widgetProps: WidgetProps): string | undefined => {
+  const { backToSignInLink, signOutLink } = widgetProps;
+  return backToSignInLink || signOutLink;
+};
+
+export const getBaseUrl = (widgetProps: WidgetProps): string | undefined => {
```
Comment (glenfannin-okta) [inline src/v3/src/util/settingsUtils.ts:82]: For reference: https://github.com/okta/okta-signin-widget/blob/21b98876539d6c48a7108a44de337027085f41d7/src/models/Settings.ts#L393-L404

File: src/v3/src/util/settingsUtils.ts [hunk 6]
```diff
@@ -73,3 +73,25 @@ export const getLanguageCode = (widgetProps: WidgetProps): LanguageCode => {
 
   return (supportedLanguages[supportedPos] ?? config.defaultLanguage) as LanguageCode;
 };
+
+export const getBackToSignInUri = (widgetProps: WidgetProps): string | undefined => {
```
Comment (glenfannin-okta) [inline src/v3/src/util/settingsUtils.ts:77]: For reference: https://github.com/okta/okta-signin-widget/blob/21b98876539d6c48a7108a44de337027085f41d7/src/models/Settings.ts#L182-L186

File: src/v3/test/integration/error-authenticator-enrollment-not-allowed.test.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+describe('error-authenticator-enrollment-not-allowed', () => {
+  it('should render form', async () => {
+    const { container, findByText } = await setup({ mockResponse });
+    await findByText(/Authenticator enrollment is not allowed at this time./);
+    expect(container).toMatchSnapshot();
+  });
+
+  // TODO: OKTA-528448 : Re-enable once this auth-js bug is fixed
+  it.skip('should send correct payload when clicking back to sign in link', async () => {
```
Comment (glenfannin-okta) [inline src/v3/test/integration/error-authenticator-enrollment-not-allowed.test.tsx:25]: Skipping this test until auth-js fix can be implemented.

File: src/v3/src/transformer/terminal/transformTerminalTransaction.ts [hunk 8]
```diff
@@ -237,6 +241,10 @@ export const transformTerminalTransaction = (
 
   const { messages } = transaction;
 
+  if (containsMessageKey(TERMINAL_KEY.SESSION_EXPIRED, messages)) {
+    authClient?.transactionManager.clear();
+  }
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/terminal/transformTerminalTransaction.ts:246]: For reference: https://github.com/okta/okta-signin-widget/blob/21b98876539d6c48a7108a44de337027085f41d7/src/v2/controllers/FormController.ts#L367-L370

File: src/v3/src/types/schema.ts [hunk 9]
```diff
@@ -277,6 +277,7 @@ export interface LinkElement extends UISchemaElement {
     label: string;
     href?: string;
     dataSe?: string;
+    onClick?: () => Promise<void>;
```
Comment (shawyu-okta) [inline src/v3/src/types/schema.ts:280]: is the return type necessary? i wonder if it could be `() => unknown`

File: src/v3/src/components/Widget/index.tsx [hunk 10]
```diff
@@ -100,6 +100,37 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
...
+    return null;
+  };
+
+  const bootstrap = useCallback(async () => {
+    try {
+      const transaction = await authClient.idx.start({
+        stateHandle: stateToken,
+      });
+
+      setAuthApiError(undefined);
```
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:122]: can we use `null` instead?

**Review by shawyu-okta — APPROVED** [review]
